[FLOC-3514] YAML configuration of benchmark scenarios, operations, and metrics

jongiddy opened the pull request:
It is useful to perform the same benchmark for a series of parameters (e.g. measured with 10, 20, 30 request/sec). Command line options for benchmarking make it difficult to parametrize the various options. 

This PR allows the scenarios, operations, and metrics provided to the benchmarking script to be read from a YAML file, allowing each one to contain additional fields for parameterization.

An example YAML file is added, in which the `wait` operation has been parameterized to wait 10 seconds.

This can be tested by running:
```
./benchmark --control ${FLOCKER_ACCEPTANCE_CONTROL_NODE} --certs ${FLOCKER_ACCEPTANCE_API_CERTIFICATES_PATH} --operation wait-10 --metric cputime
```

The options to the command now name a scenario, operation, or metric in the YAML file, which maps to a type, which selects the class to use, and a set of keyword parameters to that class.  Classes are also provided with reactor and control service parameters.

The creation of the classes is moved from `script.py` to `_driver.py` to allow creation of the class with the reactor parameter.

The above command returns output as before, but with the operation field containing the parameters:
```
{
  "scenario": {
    "type": "no-load"
  }, 
  "timestamp": "2015-11-30T12:17:11.961664", 
  "metric": {
    "type": "cputime"
  }, 
  "client": {
    "username": "jon", 
    "platform": "Linux-3.13.0-68-generic-x86_64-with-Ubuntu-14.04-trusty", 
    "working_directory": "/home/jon/projects/clusterhq/flocker-alt/benchmark", 
    "nodename": "ecuador", 
    "flocker_version": "1.8.0+84.g9943194"
  }, 
  "samples": [
    {
      "value": {
        "172.99.68.188": {
          "flocker-dataset": 1, 
          "flocker-docker-": 0, 
          "flocker-contain": 0
        }, 
        "172.99.67.245": {
          "flocker-control": 0, 
          "flocker-dataset": 0, 
          "flocker-docker-": 0, 
          "flocker-contain": 0
        }
      }, 
      "success": true
    }, 
    [...]
  ], 
  "control_service": {
    "flocker_version": "1.8.0+82.g7e996e7", 
    "host": "172.99.67.245", 
    "port": 4523
  }, 
  "operation": {
    "wait_seconds": 10, 
    "type": "wait"
  }
}
```

jongiddy commented on benchmark/script.py:
```diff
@@ -56,33 +31,48 @@ class BenchmarkOptions(Options):
+scenario_config = scenarios.get(scenario_name) or usage(
+    'No such scenario: {!r}'.format(scenario_name))
+operation_config = operations.get(operation_name) or usage(
+    'No such operations: {!r}'.format(operation_name))
```
`operations` -> `operation`

zubron commented on benchmark/operations/no_op.py:
```diff
@@ -20,14 +19,15 @@ def cleanup(self):
     An nop operation.
     """
 
     # attributes unused, but required for __init__ signature
```
We no longer need this comment.

zubron commented on benchmark/operations/read_request.py:
```diff
@@ -23,14 +23,15 @@ def cleanup(self):
     An operation to perform a read request on the control service.
     """
 
     # `clock` unused, but required for __init__ signature
```
We no longer need this comment.

zubron commented on benchmark/script.py:
```diff
@@ -95,14 +85,14 @@ class BenchmarkOptions(Options):
+    operation=operation_config.copy(),
+    metric=metric_config.copy(),
 )
 
```
Could you add a comment to explain why we choose to copy the dictionaries at this point? Either that or provide some documentation for `driver` that explains that the different `*_config` parameters will be modified.

zubron commented on benchmark/script.py:
```diff
@@ -56,33 +31,48 @@ class BenchmarkOptions(Options):
+    config = yaml.safe_load(f)
+    scenarios = config['scenarios']
+    operations = config['operations']
+    metrics = config['metrics']
```
It would be good to add validation of this configuration file. If a particular section in the file has an invalid configuration, it will result in a `KeyError` when it is used. When this error is encountered, it's unclear what the cause is. For example, if the key `type` is missing for any of the `scenarios`, `operations`, or `metrics` then the error will not be encountered until the `driver` has started.

zubron replied:
We do validation of yaml configuration elsewhere. See https://github.com/ClusterHQ/flocker/blob/eb26ab07c0902757099a7113abdef52580ec36cc/flocker/node/script.py#L169-L217

zubron commented:
Thanks for this @jongiddy. I've requested a few changes, please address the comments and then resubmit.

jongiddy commented:
Addressed all comments, and updated documentation to describe configuration file.

Minor fix to OS detection code, to ensure tests are _not_ skipped on Ubuntu.

jongiddy commented:
Also, moved configuration parsing back to main script file, which allows nicer error messages, and removes the need to copy the config dicts.

zubron commented:
This look good. Thanks for making the additional changes :) Please merge when the tests pass.
